### ISSUE TITLE
leaking file descriptors

### ISSUE THREAD

**Issue by df7cb** [oftc/node-ircws#1]
index.js seems to be leaking file descriptors:

```
lr-x------ 1 webirc webirc 64 Aug 11 20:01 0 -> /dev/null
lrwx------ 1 webirc webirc 64 Aug 11 20:01 1 -> socket:[29980715]
lrwx------ 1 webirc webirc 64 Aug 11 20:01 10 -> socket:[29980719]
lrwx------ 1 webirc webirc 64 Aug 11 20:01 11 -> socket:[29980720]
lrwx------ 1 webirc webirc 64 Aug 11 20:01 2 -> socket:[29980715]
lr-x------ 1 webirc webirc 64 Aug 11 20:01 3 -> pipe:[29980716]
l-wx------ 1 webirc webirc 64 Aug 11 20:01 4 -> pipe:[29980716]
lrwx------ 1 webirc webirc 64 Aug 11 20:01 5 -> anon_inode:[eventpoll]
lr-x------ 1 webirc webirc 64 Aug 11 20:01 6 -> pipe:[29980717]
l-wx------ 1 webirc webirc 64 Aug 11 20:01 7 -> pipe:[29980717]
lrwx------ 1 webirc webirc 64 Aug 11 20:01 8 -> anon_inode:[eventfd]
lr-x------ 1 webirc webirc 64 Aug 11 20:01 9 -> /
```

... one connect-disconnect later:

```
lr-x------ 1 webirc webirc 64 Aug 11 20:01 0 -> /dev/null
lrwx------ 1 webirc webirc 64 Aug 11 20:01 1 -> socket:[29980715]
lrwx------ 1 webirc webirc 64 Aug 11 20:01 10 -> socket:[29980719]
lrwx------ 1 webirc webirc 64 Aug 11 20:01 11 -> socket:[29980720]
lrwx------ 1 webirc webirc 64 Aug 11 20:01 12 -> socket:[29991176]
lrwx------ 1 webirc webirc 64 Aug 11 20:01 13 -> socket:[29991206]
lrwx------ 1 webirc webirc 64 Aug 11 20:01 16 -> socket:[29991214]
lrwx------ 1 webirc webirc 64 Aug 11 20:01 2 -> socket:[29980715]
lr-x------ 1 webirc webirc 64 Aug 11 20:01 3 -> pipe:[29980716]
l-wx------ 1 webirc webirc 64 Aug 11 20:01 4 -> pipe:[29980716]
lrwx------ 1 webirc webirc 64 Aug 11 20:01 5 -> anon_inode:[eventpoll]
lr-x------ 1 webirc webirc 64 Aug 11 20:01 6 -> pipe:[29980717]
l-wx------ 1 webirc webirc 64 Aug 11 20:01 7 -> pipe:[29980717]
lrwx------ 1 webirc webirc 64 Aug 11 20:01 8 -> anon_inode:[eventfd]
lr-x------ 1 webirc webirc 64 Aug 11 20:01 9 -> /
```

**Comment by tjfontaine** [comment]
Are these connects that successfully negotiated, or were they interrupted before being connected? Also knowing if that's the inbound connection or the outbound would be ideal. What version of Node.js is being used?

**Comment by df7cb** [comment]
I don't know which connections are concerned, so far I've just noticed that there's more open file descriptors than what makes sense to me. 

On dibasic, which doesn't have any users anymore, the fd list looks like this:

```
c---------   1 root     sys       38,  2 Sep  3 03:10 10
c---------   1 root     tty       59,  0 Aug 28 17:42 1
c---------   1 root     tty       59,  0 Aug 28 17:42 11
c---------   1 root     tty       59,  0 Aug 28 17:42 12
c---------   1 root     tty       59,  0 Aug 28 17:42 2
p---------   0 oftc     other          0 Aug 12 20:27 8|
p---------   0 oftc     other          0 Aug 12 20:27 9|
p---------   0 oftc     other          1 Aug 11 18:41 3|
p---------   0 oftc     other          0 Aug 11 18:41 4|
P---------   2 oftc     other          0 Aug 11 18:41 5
c---------   1 oftc     tty        6,  2 Aug 11 18:41 0
p---------   0 oftc     other          0 Aug 11 18:41 6|
p---------   0 oftc     other          0 Aug 11 18:41 7|
D---------   1 root     root           0 Jun 16 18:37 17>
s---------   0 root     root           0 Jan  1  1970 13=
s---------   0 root     root           0 Jan  1  1970 14=
s---------   0 root     root           0 Jan  1  1970 20=
s---------   0 root     root           0 Jan  1  1970 24=
s---------   0 root     root           0 Jan  1  1970 31=
s---------   0 root     root           0 Jan  1  1970 34=
s---------   0 root     root           0 Jan  1  1970 36=
s---------   0 root     root           0 Jan  1  1970 44=
s---------   0 root     root           0 Jan  1  1970 52=
s---------   0 root     root           0 Jan  1  1970 53=
```

That's with node v0.10.28 I think.

**Comment by df7cb** [comment]
The problem here seems to be that browsers tend to keep the websocket connection open, even if the corresponding tab is already closed. We properly close the IRC connection, but the websocket connection stays open until people close the browser.

plasma currently has ~400 websocket connections open, while there's only 6 IRC connections left.

**Comment by prawnsalad** [comment]
Just came across this issue and wanted to point out that these type of issues are why people move away from socket.io. It's just not that great when it comes to anything larger than a handful connections.

**Comment by dwfreed** [comment]
This appears to have been fixed by daa826e, so closing.
